File: java/src/bg/util/leto/impl/gregorian/LetoGregorianMonth.java
package bg.util.leto.impl.gregorian;

import java.util.Locale;
import java.util.Map;

import bg.util.leto.api.LetoExceptionUnrecoverable;
import bg.util.leto.api.LetoPeriodStructure;
import bg.util.leto.api.LetoPeriodType;
import bg.util.leto.base.LetoPeriodStructureBean;
import bg.util.leto.impl.LocaleStringId;
import bg.util.leto.impl.LocaleStrings;

public class LetoGregorianMonth implements LetoPeriodStructure {
    
    
    private LetoPeriodStructureBean mBean = null;
    private LocaleStringId mNameTranslationId = null;
    private LetoPeriodType mPeriodType = null;
    private Map<LetoPeriodType, Long> mTotalLengthInPeriodTypes = null;
    
    public LetoGregorianMonth(LetoPeriodStructureBean bean, LocaleStringId translatedName) {
        mBean = bean;
        mNameTranslationId = translatedName;
    }

    @Override
    public LetoPeriodType getPeriodType() {
        if (mPeriodType == null) {
            return mBean.getPeriodType();
        } else {
            return mPeriodType;
        }
    }

    @Override
    public void setPeriodType(LetoPeriodType period) throws LetoExceptionUnrecoverable {
        mPeriodType = period;
    }

    @Override
    public long getTotalLengthInDays() {
        return mBean.getTotalLengthInDays();
    }

    @Override
    public LetoPeriodStructure[] getSubPeriods() {
        return mBean.getSubPeriods();
    }
    
    public void setTotalLengthInPeriodTypes(Map<LetoPeriodType, Long> lengthsInPeriodTypes) {
        mTotalLengthInPeriodTypes = lengthsInPeriodTypes;
    }

    @Override
    public long getTotalLengthInPeriodTypes(LetoPeriodType periodType) {
        if (periodType == getPeriodType()) {
            return 1;
        }
        if (mTotalLengthInPeriodTypes == null) {
            mBean.getTotalLengthInPeriodTypes(periodType);
        }
        Long countLong = mTotalLengthInPeriodTypes.get(periodType);
        if (countLong == null) {
            mBean.getTotalLengthInPeriodTypes(periodType);
        }
        return countLong.longValue();
    }

    @Override
    public String getName() {
        return getName(Locale.ENGLISH);
    }
    
    @Override
    public String getName(Locale locale) {
        return LocaleStrings.get(mNameTranslationId, locale, mBean.getName(locale));
    }

    @Override
    public Map<Locale, String> getNameTranslations() {
        Map<Locale, String> translations = LocaleStrings.get(mNameTranslationId);
        return translations;
    }

}

File: javascript/bg/util/leto/impl/gregorian/LetoGregorianMonth.js

function LetoGregorianMonth(/*LetoPeriodStructureBean*/ bean, /*LocaleStringId*/ translatedName) {
    this.mBean = bean;                         // LetoPeriodStructureBean
    this.mNameTranslationId = translatedName;  // LocaleStringId

    this.mPeriodType = null;                   // LetoPeriodType 
    this.mTotalLengthInPeriodTypes = null;     // Map<LetoPeriodType, Long> mTotalLengthInPeriodTypes
    

    this.getPeriodType = function () {
        if (mPeriodType == null) {
            return mBean.getPeriodType();
        } else {
            return mPeriodType;
        }
    }

    this.setPeriodType = function (period) {
        mPeriodType = period;
    }

    this.getTotalLengthInDays = function () {
        return mBean.getTotalLengthInDays();
    }

    this.getSubPeriods = function () {
        return mBean.getSubPeriods();
    }
    
    this.setTotalLengthInPeriodTypes = function (lengthsInPeriodTypes) {
        mTotalLengthInPeriodTypes = lengthsInPeriodTypes;
    }

    this.getTotalLengthInPeriodTypes = function (periodType) {
        if (periodType == getPeriodType()) {
            return 1;
        }
        if (mTotalLengthInPeriodTypes == null) {
            mBean.getTotalLengthInPeriodTypes(periodType);
        }
        var countLong = mTotalLengthInPeriodTypes.get(periodType);
        if (countLong == null) {
            mBean.getTotalLengthInPeriodTypes(periodType);
        }
        return countLong.longValue();
    }
    
    this.getName = function (locale) {
        if (locale == null) {
           locale = "en";
        } 
        return LocaleStrings.get(mNameTranslationId, locale, mBean.getName(locale));
    }

    this.getNameTranslations = function () {
        var translations = LocaleStrings.get(mNameTranslationId);
        return translations;
    }

}

if (module != null && module.exports != null)  {
  module.exports = LetoGregorianMonth
}

